libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProvider.kt

```diff
@@ -0,0 +1,116 @@
+ */
+class GuidanceViewImageProvider() {
+
+    companion object {
+        private val USER_AGENT_KEY = "User-Agent"
+        private val USER_AGENT_VALUE = "MapboxJava/"
+    }
+
+    /**
+     * This is added a solution to bypass the issue with Picasso where [com.squareup.picasso.Target]
+     * get's garbage collected and guidance view is hence not rendered. Found the solution here
+     * https://stackoverflow.com/questions/24180805/onbitmaploaded-of-target-object-not-called-on-first-load#answers
+     */
+    private val targets: MutableList<Target> = mutableListOf()
+    private val okHttpClient = OkHttpClient.Builder().addInterceptor { chain: Interceptor.Chain ->
+        chain.proceed(
+            chain.request().newBuilder().addHeader(USER_AGENT_KEY, USER_AGENT_VALUE).build()
+        )
+    }.build()
+
+    /**
+     * The API reads the bannerInstruction and renders a guidance view if one is available
+     * @param bannerInstructions [BannerInstructions]
+     * @param context [Context]
+     * @param callback [OnGuidanceImageDownload] Callback that is triggered based on appropriate state of image downloading
+     */
+    fun renderGuidanceView(
+        bannerInstructions: BannerInstructions,
+        context: Context,
+        callback: OnGuidanceImageDownload
+    ) {
+        val target = object : Target {
+            override fun onPrepareLoad(placeHolderDrawable: Drawable?) {
+            }
+
+            override fun onBitmapFailed(e: Exception?, errorDrawable: Drawable?) {
+                targets.clear()
+                callback.onFailure(e?.message)
+            }
+
+            override fun onBitmapLoaded(bitmap: Bitmap?, from: Picasso.LoadedFrom?) {
+                targets.clear()
+                ifNonNull(bitmap) { b ->
+                    callback.onGuidanceImageReady(b)
+                } ?: callback.onFailure("Something went wrong. Bitmap not received")
+            }
+        }
+        val bannerView = bannerInstructions.view()
+        ifNonNull(bannerView) { view ->
+            val bannerComponents = view.components()
+            ifNonNull(bannerComponents) { components ->
+                components.forEachIndexed { _, component ->
+                    component.takeIf { c -> c.type() == BannerComponents.GUIDANCE_VIEW }?.let {
+                        targets.add(target)
+                        Picasso.Builder(context).downloader(OkHttp3Downloader(okHttpClient))
+                            .build()
+                            .load(it.imageUrl())
+                            .into(target)
+                    }
+                }
+            } ?: callback.onNoGuidanceImageUrl()
+        } ?: callback.onNoGuidanceImageUrl()
+    }
+
+    /**
+     * The API allows you to cancel the rendering of guidance view.
+     * @param context Context
+     */
+    fun cancelRender(context: Context) {
+        if (targets.isNotEmpty()) {
+            Picasso.Builder(context).build().cancelRequest(targets[0])
```

Comment:
we need to cancel all, not only the first one

Reply:
The list of `targets` can never have more than 1 target, because a single banner instruction cannot have more than 1 guidance view. In case of subsequent banner instructions containing guidance view, by the time the 2nd banner instruction is emitted the first one would have already been displayed and the target would have been cleared. Hence this will work fine.
cc @LukasPaczos

Reply:
This might not be the case if the picasso request doesn't manage to return before the next banner instructions are provided (due to poor connectivity or anything else). Hypothetically, the user's app logic might also cache instructions and request more than one image at a time. It's more of an egde case, but it'd be great to take it into account.